halo2_proofs/src/dev.rs

```diff
@@ -97,16 +97,118 @@ impl Region {
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (Self::Unassigned, Self::Unassigned) => true,
+            (Self::Assigned(a), Self::Assigned(b)) => a == b,
+            (Self::Rational(a, b), Self::Rational(c, d)) => *a * d == *b * c,
+            (Self::Poison(a), Self::Poison(b)) => a == b,
+            _ => false,
+        }
+    }
+}
+
+impl<F: Group + Field> CellValue<F> {
+    /// Returns the numerator.
+    pub fn numerator(&self) -> Option<F> {
+        match self {
+            Self::Rational(numerator, _) => Some(*numerator),
+            _ => None,
+        }
+    }
+
+    /// Returns the denominator
+    pub fn denominator(&self) -> Option<F> {
+        match self {
+            Self::Rational(_, denominator) => Some(*denominator),
+            _ => None,
+        }
+    }
+}
+
+impl<F: Group + Field> From<Assigned<F>> for CellValue<F> {
+    fn from(value: Assigned<F>) -> Self {
+        match value {
+            Assigned::Zero => CellValue::Unassigned,
+            Assigned::Trivial(value) => CellValue::Assigned(value),
+            Assigned::Rational(numerator, denominator) => {
+                CellValue::Rational(numerator, denominator)
+            }
+        }
+    }
+}
+
+fn calculate_assigned_values<F: Group + Field>(
+    cell_values: &mut [CellValue<F>],
+    inv_denoms: &[Option<F>],
+) {
+    assert_eq!(inv_denoms.len(), cell_values.len());
+    for (value, inv_den) in cell_values.iter_mut().zip(inv_denoms.iter()) {
+        // if numerator and denominator exist, calculate the assigned value
+        // otherwise, return the original CellValue
+        *value = match value {
+            CellValue::Rational(numerator, _) => CellValue::Assigned(*numerator * inv_den.unwrap()),
+            _ => *value,
+        };
+    }
+}
+
+fn batch_invert_cellvalues<F: Field + Group>(cell_values: &mut [Vec<CellValue<F>>]) {
+    let mut denominators: Vec<_> = cell_values
+        .par_iter()
+        .map(|f| {
+            f.iter()
+                .map(|value| value.denominator())
+                .collect::<Vec<_>>()
+        })
+        .collect();
+
+    let mut mut_denominators = denominators
+        .iter_mut()
+        .flat_map(|f| {
+            f.iter_mut()
+                // If the denominator is trivial, we can skip it, reducing the
+                // size of the batch inversion.
+                .filter_map(|d| d.as_mut())
+        })
+        .collect::<Vec<_>>();
+
+    // copy from halo2-gpu/halo2_proofs/src/poly.rs: `fn batch_invert_assigned_par`
+    let num_threads = rayon::current_num_threads();
+    let chunk_size = (mut_denominators.len() + num_threads - 1) / num_threads;
+    let mut chunks = vec![];
+    while mut_denominators.len() > chunk_size {
+        let chunk = mut_denominators.split_off(mut_denominators.len() - chunk_size);
+        chunks.push(chunk);
+    }
+    chunks.push(mut_denominators);
+    rayon::scope(|scope| {
+        for chunk in chunks {
```

Comment:
why don't we use `mut_denominators.chunks_mut(chunk_size)`?

Reply:
it seems that &mut [&mut F] does not satisfy the trait bound to batch_invert()
so I followed [the same approach here as we did in halo2-gpu](https://github.com/scroll-tech/halo2-gpu/blob/728cf00b65d493472f7310f75c74b44e5e51b98b/halo2_proofs/src/poly.rs#L257-L273)

Reply:
yes, you're right.